### PR TITLE
Added max and avg pool, alongside tests and docs

### PR DESCRIPTION
**Pull request by Benjamin-Walker**
I currently use einops.rearrange so that the pooling layers and conv layers take the same tensor format as input, BCHW. For some reason lax.conv_general_dilated and lax.reduce_window are BCHW and BHWC respectively. Can do this not using einops if you would prefer to not have the extra dependency.

### CODE REVIEW

File: equinox/nn/pool.py [hunk 1]
```diff
@@ -0,0 +1,201 @@
...
+            x = x / np.prod(self.kernel_size)
+
+        x = jnp.squeeze(x, axis=0)
+        x = einops.rearrange(x, '... a -> a ...')
+        return x
+
+
+class AvgPool1D(Pool):
+    """As [`equinox.nn.Pool`][] with `operation=lax.add` and
+    `num_spatial_dims=1`."""
```
Comment (patrick-kidger) [inline equinox/nn/pool.py:86]: I can tell you've copied the style of the convolutional layers here, but I think this docstring is a bit opaque. Something like "downsamples by taking an average over a sliding window" might be a helpful one-sentence summary.

**Review by patrick-kidger — COMMENTED** [review]
Yes, I'd prefer to avoid the additional dependency on einops.

Can you add some tests for the backward pass? As per #66 it seems like that's a potential issue we may need to try and resolve.

**Comment by patrick-kidger** [comment]
Also: I think the new pooling documentation needs to be added to `mkdocs.yml`.

File: equinox/nn/pool.py [hunk 2]
```diff
@@ -0,0 +1,201 @@
...
+    padding: Union[str, Sequence[Tuple[int, int]]] = static_field()
+    init: float = static_field()
+
+    def __init__(
+        self,
+        operation: Callable[[Array], Array],
+        num_spatial_dims: int,
+        kernel_size: Union[int, Sequence[int]],
+        strides: Sequence[int] = None,
+        padding: Union[str, Sequence[Tuple[int, int]]] = "VALID",
```
[the remark this answers is not in the document]
Comment (Benjamin-Walker) [reply equinox/nn/pool.py:26]: Sure, I will also include Sequence[Tuple[int, int]], as you can't have 'SAME' padding with only int and Sequence[int].

File: equinox/nn/pool.py [hunk 3]
```diff
@@ -0,0 +1,238 @@
...
+import numpy as np
+
+from ..custom_types import Array
+from ..module import Module, static_field
+
+
+class Pool(Module):
+    """ General N-dimensional downsampling over a sliding window. """
+
+    operation: Callable[[Array], Array] = static_field()
```
Comment (patrick-kidger) [inline equinox/nn/pool.py:16]: This shouldn't be static. (In principle the operation may be learnt.) Also I think the annotation should be `Callable[[Array, Array], Array]`.

File: equinox/nn/pool.py [hunk 4]
```diff
@@ -0,0 +1,238 @@
...
+    num_spatial_dims: int = static_field()
+    kernel_size: Union[int, Sequence[int]] = static_field()
+    stride: Union[int, Sequence[int]] = static_field()
+    padding: Union[int, Sequence[int]] = static_field()
+    init: float = static_field()
+
+    def __init__(
+        self,
+        operation: Callable[[Array], Array],
```
Comment (patrick-kidger) [inline equinox/nn/pool.py:25]: Likewise annotation here

File: equinox/nn/pool.py [hunk 5]
```diff
@@ -0,0 +1,238 @@
...
+        - `stride`: The stride of the convolution.
+        - `padding`: The amount of padding to apply before and after each
+            spatial dimension.
+        """
+        super().__init__(**kwargs)
+
+        self.operation = operation
+        self.init = 0.0
+        if operation is lax.max:
+            self.init = -jnp.inf
```
Comment (patrick-kidger) [inline equinox/nn/pool.py:45]: Make `init` an input, don't try to guess based on special values. Essentially `Pool` should make sense as a standalone operation.

File: equinox/nn/pool.py [hunk 6]
```diff
@@ -0,0 +1,238 @@
...
+        else:
+            raise ValueError(
+                "`stride` must either be an int or tuple of length "
+                f"{num_spatial_dims} containing ints."
+            )
+
+        if isinstance(padding, int):
+            self.padding = tuple((padding, padding) for _ in range(num_spatial_dims))
+        elif isinstance(padding, Sequence) and \
+            all(isinstance(element, tuple) for element in padding):
```
Comment (patrick-kidger) [inline equinox/nn/pool.py:71]: `isinstance(element, tuple)`
to
`isinstance(element, Sequence)`

File: tests/test_nn.py [hunk 7]
```diff
@@ -6,6 +6,8 @@
 import jax.random as jrandom
 import pytest
 
+import sys
+sys.path.append('/Users/benwalker/PycharmProjects/equinox')
```
[the remark this answers is not in the document]
Comment (Benjamin-Walker) [reply tests/test_nn.py:10]: Hahahah oops, didn't mean to leave that one in

File: equinox/nn/pool.py [hunk 8]
```diff
@@ -0,0 +1,245 @@
...
+    operation: Callable[[Array, Array], Array]
+    num_spatial_dims: int = static_field()
+    kernel_size: Union[int, Sequence[int]] = static_field()
+    stride: Union[int, Sequence[int]] = static_field()
+    padding: Union[int, Sequence[int]] = static_field()
+    init: float = static_field()
+
+    def __init__(
+        self,
+        init: int,
```
Comment (patrick-kidger) [inline equinox/nn/pool.py:24]: Nit: not necessarily an integer! e.g. `jnp.NINF` is a float.

Admittedly I'm not sure what the correct annotation is here. `Union[int, float]` maybe? Probably `Array`s are valid too?

**Review by patrick-kidger — COMMENTED** [review]
Yep, good thought; add that somewhere in the documentation for `Pool`. Preferably with a few references to the relevant JAX issues etc.

**Comment by Benjamin-Walker** [comment]
Fixed the remaining typesetting mistake

**Comment by Benjamin-Walker** [comment]
I think it was actually 3.10 where they introduced using | instead of Union. Think I have squashed all the commits together, let me know if it hasn't worked.

File: tests/test_nn.py [hunk 9]
```diff
@@ -653,3 +655,116 @@ def get_weights(m):
...
+    x = jnp.arange(64, dtype=jnp.float32).reshape(1, 4, 4, 4)
+    grad_fn = jax.value_and_grad(max_pool_mean)
+
+    grad_fn(x)
+
+
+def test_poolnetworkbackprop(getkey):
+    class CNN(eqx.Module):
+        conv_layer: list[Union[eqx.nn.Conv2d, eqx.nn.MaxPool2D]]
+        linear_layers: list[eqx.nn.Linear]
```
Comment (patrick-kidger) [inline tests/test_nn.py:742]: This test will still fail because `list[...]` is 3.9+; you need `typing.List[...]`.

Comment (Benjamin-Walker) [reply tests/test_nn.py]: Fixed now, sorry about all the little mistakes!

Comment (patrick-kidger) [reply tests/test_nn.py]: @Benjamin-Walker still failing!
This will go a lot faster [if you install pre-commit](https://github.com/patrick-kidger/equinox/blob/main/CONTRIBUTING.md).

Comment (Benjamin-Walker) [reply tests/test_nn.py]: Sorry I looked online for how to run pre-commits locally but I should have just read the docs. Hopefully all fine now, all the tests pass locally.
